src/services/Batcher.ts

```diff
@@ -0,0 +1,102 @@
+import { MaybePromise } from '@/types'
+
+export type BatchCallbackLookup<V, R> = (value: V) => MaybePromise<R>
+export type BatchCallback<V, R> = (values: V[]) => MaybePromise<Map<V, R>> | BatchCallbackLookup<V, R>
+export type BatchOptions = {
+  maxBatchSize?: number,
+  maxCycles?: number,
+}
+
+type Batch<V, R> = Map<V, BatchPromise<R>>
+type BatchPromise<T> = {
+  resolve: (value: T) => void,
+  reject: (reason?: unknown) => void,
+}
+
+export class Batcher<V, R> {
+  private readonly queue: Batch<V, R> = new Map()
+  private timeout: ReturnType<typeof setTimeout> | null = null
+  private readonly callback: BatchCallback<V, R>
+  private readonly options: BatchOptions
+  private readonly loops: number = 0
+
+  public constructor(callback: BatchCallback<V, R>, options: BatchOptions = {}) {
+    this.callback = callback
+    this.options = options
+  }
+
+  public batch(value: V): Promise<R> {
+
+    const response = new Promise<R>((resolve, reject) => {
+      this.queue.set(value, { resolve, reject })
```

Comment:
So since `batch` takes just one argument instead of spreading unknown arguments we're encouraging callbacks that require multiple arguments to be passed as arrays. TIL you can use objects as Map keys but is there a better/clearer pattern here? I was struggling to understand the single call arguments => queue key connection when reading through this

Reply:
I wasn't even considering callbacks with multiple arguments. We can update these easily to support multiple args. Good catch

Reply:
The queue is keyed on the value you're batching. Every time you call batch a new key is added to the queue with an unresolved promise and you get that promise back as the return value from the batch call. 

The callback accepts the value (`V`) that was batched and must return `R`. `R` is used to resolve the promise. 

This will all work the same way for multiple args. But I think the queue will need to be keyed with the batch value stringified. Otherwise if you batch the same value twice you might end up with an unresolved promise that just waits forever.

Reply:
@znicholasbrown thinking more about this. I think a single argument is actually preferred. The batch method is for batching operations, not method calls. If you have an operation you want to perform multiple times you can init a batcher with a callback and then call batch with a single argument. If your operation needs multiple values then you're able to pass an object or an array into the batcher. How the callback processes those values is an implementation detail. 

But I do need to make sure if a value already exists in the map that a promise isn't lost.